proposals/0000-splice-imports.rst

```diff
@@ -0,0 +1,1058 @@
+the syntax of ``splice`` and ``quote`` imports without obliging the whole module
+to be level-correct.
+
+
+Syntax for imports
+##################
+
+Under ``ExplicitLevelImports``, the syntax for imports becomes::
+
+  importdecl :: { LImportDecl GhcPs }
+     : 'import' maybe_src maybe_safe optsplice optqualified maybe_pkg modid optqualified maybeas maybeimpspec
+
+  optsplice :: { LImportStage }
+     : 'splice' { SpliceStage }
+     | 'quote'  { QuoteStage  }
+     |          { NormalStage }
+
+
+The ``splice`` or ``quote`` keyword appears before the ``qualified`` keyword but after ``SOURCE``
+and ``SAFE`` pragmas.
+
+
+Name resolution and ambiguous variables
+#######################################
+
+Name resolution ("renaming") does not take account of the level at which an
+identifier was imported when disambiguating ambiguous names, even though this is
+sometimes more conservative than necessary.  For example, the following program
+is rejected::
+
+  {-# LANGUAGE ExplicitLevelImports #-}
+
+  import A ( x )
+  import splice B ( x )
+
+  foo = $( x ) x
+
+In this case, there is in principle no ambiguity because ``A.x`` isn't allowed
+to be used in the top-level splice, and ``B.x`` isn't allowed to be used outside
+the splice.  Thus the only disambiguation that will pass the type-checker is::
+
+  foo = $( B.x ) A.x
+
+We choose to reject this disambiguation to keep the design simple and prevent
+any confusion about what is in scope. This position is conservative, and can be
+relaxed in the future if more flexibility appears worthwhile.
+
+A positive consequence of the current design is that if a program is accepted
+with ``ExplicitLevelImports``, it will be accepted after erasing all
+``splice``/``quote`` keywords and using ``ImplicitStagePersistence`` instead of
+``ExplicitLevelImports``.
+
+
+Exports
+#######
+
+Under ``NoImplicitStagePersistence``, modules may export bindings only if they
+are available at level 0.
+
+For example, the following is rejected::
+
+  {-# LANGUAGE ExplicitLevelImports #-}
+
+  module M (oops) where  -- Error: oops imported at level -1 but used at level 0
+    import splice N ( oops )
+
+
+Class instance resolution
+#########################
+
+Class instances carry a level, much like identifiers, and must be used at the
+correct level.  This will be enforced by the type-checker under
+``NoImplicitStagePersistence``:
+
+* Instance resolution views the set of instances from all imports together and thus
+  instances from normal and splice imports must agree with each other.
+
+* After instance resolution has selected an instance, it is checked which levels
+  the instance is available at and an error is raised if the instance is not available
+  at the correct level.
+
+This design for instances mirrors the situation for name resolution. As with
+ambiguous names, it would in principle be possible for the type-checker to make
+use of level information to accept more programs, but this seems like an
```

Comment:
@adamgundry says that he will reply to this later.

Reply:
I think @aspiwack is right here. The proposed design ensures coherence across stages (or at least, permits no more incoherence than GHC allows anyway!). But if the type-checker were to solve constraints based on availability of instances, it could end up using different instances for the same constraint at different stages, thereby violating coherence. So in fact the proposal could be more explicit that using level information in the constraint solver is a bad idea.

Reply:
Ideally this is something we can revisit later. I would expect that it shouldn't be so hard to implement if type contexts and other such things are rather independent for levels --- indeed, we always have the option to basically treat each level almost like an independent module.

This proposal is about the language changes, which should --- and I believe does --- admit many implementation strategies.